File: src/main/java/com/github/xumeng/java/problems/easy/problem0290/Solution0290.java
package com.github.xumeng.java.problems.easy.problem0290;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution0290 {

    public boolean wordPattern(String pattern, String s) {
        String[] stringElements = s.split(" ");
        String[] patternElements = pattern.split("");
        if (stringElements.length != patternElements.length) {
            return false;
        }
        Map<String, String> map = new HashMap<>();
        List<String> mapValueElements = new ArrayList<>();
        for (int i = 0; i < stringElements.length; i++) {
            String patternElement = patternElements[i];
            String stringElement = stringElements[i];
            String target = map.get(patternElement);
            if (target == null) {
                if (mapValueElements.contains(stringElement)) {
                    return false;
                } else {
                    map.put(patternElement, stringElement);
                    mapValueElements.add(stringElement);
                }
            } else {
                if (!target.equals(stringElement)) {
                    return false;
                }
            }
        }
        return true;
    }

}

File: src/main/kotlin/com/github/xumeng/kotlin/problems/easy/problem0290/Solution0290.kt
package com.github.xumeng.kotlin.problems.easy.problem0290

class Solution0290 {

    fun wordPattern(pattern: String, s: String): Boolean {
        val stringElements: List<String> = s.split(" ")
        val patternElements: MutableList<String> = mutableListOf()
        for (i in pattern.indices) {
            patternElements.add(pattern[i].toString())
        }
        if (stringElements.size != patternElements.size) {
            return false
        }
        val map: MutableMap<String, String> = mutableMapOf()
        val mapValueElement: MutableList<String> = mutableListOf()
        for (i in patternElements.indices) {
            val patternElement = patternElements[i]
            val stringElement = stringElements[i]
            val target: String? = map[patternElement]
            if (target == null) {
                if (mapValueElement.contains(stringElement)) {
                    return false
                } else {
                    map[patternElement] = stringElement
                    mapValueElement.add(stringElement)
                }
            } else {
                if (target == stringElement) {
                    continue
                } else {
                    return false
                }
            }
        }
        return true
    }

}